### PR TITLE
Implement <resize><drawContents>

### LINKED ISSUE #1328: Feature request: <resize><drawContents>no</drawContents> ... </resize>
With compositing on resizing a Chromium window on an aging intel IGPU is often laggy experience, with a high polling rate gaming mouse on Wayland it's horrible. Only on Hyprland it's not, maybe because of some wlroots ENVs I have on there, I don't yet know, but in KDE, Gnome, labwc it is. (Maybe the trick is that the border I'm dragging gets more CPU time on Hyprland and also maybe they are doing a trick with window content warping for resize, which is a fancy way of doing the same thing I'm asking for here in this post) 
Openbox has a feature that helps sidestepping this problem, if 
```
 <resize>
    <drawContents>no</drawContents>
...
  </resize>
```
is added to the rc.xml the system doesn't have to struggle redrawing everything, battery is saved on a laptop etc etc..
Please consider adding this, if you feel like it someday ^

Thank you (and contributors as well) for developing labwc, it's already so much better than openbox in some fundamental ways. (window aero snap without janky half-assed workarounds and the horrible titlebar seperator that ruins consistent theming is not in labwc at all!)

### PR DESCRIPTION
**Pull request by tokyo4j**
Closes #1328.

https://github.com/labwc/labwc/assets/22029524/e59a08f6-e3d1-4806-afb1-d95e3d4e8c74

The theme for the frame shown to indicate resized window when `<resize><drawcontents>` is set to no is configured with:
- `resize-frame.width`
- `resize-frame.color`

But I'm not sure about these names. Perhaps `resize.frame.{width,color}` or other names are better.

I initially implemented the frame as a combination of multi_rect and scene_rect like I did for snapping overlay, but I removed it for simplicity (we need to add `resize-frame.{bg,border}.enabled` and disable translucent scene_rect for pixman renderer).
If anyone wants it, I can move it back from [here](https://github.com/tokyo4j/labwc/tree/drawcontents-first).

The first commit adds `zero_array()` macro in `theme.c` to shorten the lengthy `memcpy()` calls like `memset(theme->snapping_overlay_region.border_color, 0, sizeof(theme->snapping_overlay_region.border_color))`.

### CODE REVIEW

File: include/resize-frame.h [hunk 1]
```diff
@@ -0,0 +1,11 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+#ifndef LABWC_RESIZE_FRAME_H
+#define LABWC_RESIZE_FRAME_H
+
+#include "view.h"
+#include <wlr/util/box.h>
+
+void resize_frame_update(struct view *view, struct wlr_box new_geo);
+void resize_frame_finish(struct view *view);
```
Comment (Consolatis) [inline include/resize-frame.h:9]: `struct view` could be forward declared rather than including `view.h`

**Comment by Consolatis** [comment]
Diff LGTM.

> But I'm not sure about these names. Perhaps resize.frame.{width,color} or other names are better.

Agree, personally I also prefer `resize.frame.{width,color}`.

If openbox provides a setting for the feature (and the theme vars) we should use whatever openbox uses though, I didn't look into it yet.

**Comment by johanmalm** [comment]
I prefer thickness to width because width has multiple meanings (i.e. also width of window)

**Comment by tokyo4j** [comment]
> Agree, personally I also prefer resize.frame.{width,color}.

I don't have a strong preference here. I used `resize-frame` because I thought `frame` might be a bit vague so I defined `resize-frame` as a proper noun. But it won't make any big difference.

I'll change it to `resize.frame`. 

> If openbox provides a setting for the feature (and the theme vars) we should use whatever openbox uses though, I didn't look into it yet.

Openbox doesn't help here because it implements `<resize><drawContents>` differently and there's no theme entries for it. If we can implement it in the same way that doesn't hurt our codebase, switching to it would be another option.

https://github.com/labwc/labwc/assets/22029524/43f81fad-e88d-4373-9474-d4363f66879d

> Somewhat related: the preview outlines (osd_update_preview_outlines) and this feels like kind of the same thing. Maybe the preview outlines should be configurable and let them both be set from the same options?

I added `osd.window-switcher.preview.border.{width,color}` in #1715. But if we follow its names here, we should use `resize.frame.border.{width,color}`.

> I prefer thickness to width because width has multiple meanings (i.e. also width of window)

Agree. I think `resize.frame.thickness` or `resize.frame.border.width` would be better for clarity. 

So, my current candidates are:
1. `resize.frame.border.{width,color}`
2. `resize.frame.{thickness,color}`

Any opinions?

File: include/resize-frame.h (hunk 1 above)
Comment (tokyo4j) [reply include/resize-frame.h:10]: Thanks. Fixed.

**Comment by Consolatis** [comment]
> So, my current candidates are:
> 
> 1. `resize.frame.border.{width,color}`
> 2. `resize.frame.{thickness,color}`
> 
> Any opinions?

With 1 we could add the openbox behavior in the future if we want to (e.g. drawing a solid rect behind the window contents on resize) and then use something like `resize.frame.color` to define its color. On the other side we could also use `resize.frame.bg.color` there. That would then be a bit weird to have `resize.frame.color` be the border color though.

So I'd prefer `1` of the two prefixes. Regarding `width` vs `thickness` I don't really have an opinion.

**Comment by tokyo4j** [comment]
> With 1 we could add the openbox behavior in the future if we want to (e.g. drawing a solid rect behind the window contents on resize) and then use something like `resize.frame.color` to define its color. On the other side we could also use `resize.frame.bg.color` there. That would then be a bit weird to have `resize.frame.color` be the border color though.

> So I'd prefer 1 of the two prefixes. Regarding width vs thickness I don't really have an opinion.

If we consider the case we implement Openbox's behavior in the future, I don't think `resize.frame.bg` is a right name there. I named `resize.frame` with the look of multi-rect in mind, and "background of frame" doesn't make sense to me (it may be just my English issue though).

When we implement the Openbox's behavior, I think a possible configuration is like this:

```xml
<!-- no is equal to frame(?) -->
<resize><drawContents>yes|no|frame|fill</drawContents></resize>
```

```
resize.frame.thickness: 1
resize.frame.color: #dddda6,#000000,#dddda6
resize.fill.color: #dddda6
```

So I now prefer `resize.frame.{thickness,color}`.

**Comment by tokyo4j** [comment]
The last thing I'm wondering is `frame` vs `outlines`. Perhaps `outlines` is better if we follow `<windowSwitcher outlines="yes">`.

BTW I'm now regretting I named `window-switcher.preview.border.{width,color}` as I didn't know the outlines can be drawn even with `<windowSwitcher preview="no" />`. I think we can change it to `window-switcher.outlines.{thickness,color}`.

**Comment by tokyo4j** [comment]
I squashed the fixup commit and updated the commit message.
This PR is ready now, but I'll wait for LTGM from @Consolatis just in case. There's no need to hurry.

**Comment by tokyo4j** [comment]
Regarding to https://github.com/labwc/labwc/issues/1328#issuecomment-2142565270, I think applying the same behavior while moving window with a configuration like `<move><drawContents>` is also worth implementing to reduce the size of the updated area.

In fvwm, they're configured with:
```
# Draw a frame on resize and move
Style * ResizeOutline
OpaqueMoveSize 0

# Let clients draw their content on resize and move
Style * ResizeOpaque
OpaqueMoveSize unlimited 
```

**Comment by johanmalm** [comment]
LGTM - except we ought to add the config option in `docs/rc.xml.all` too.

Have tested and it works well.

Nice contribution.

**Comment by Consolatis** [comment]
Had a quick test again and noticed two things:
- If the resize popup is enabled it feels a bit weird that it doesn't update on movement (neither in value nor in position).
- Sometimes the border is a bit hard to see, especially on white application backgrounds.

Do we want to fix both of these within this PR?

**Comment by tokyo4j** [comment]
>  - If the resize popup is enabled it feels a bit weird that it doesn't update on movement (neither in value nor in position).

I fixed this in the fixup commit.

> - Sometimes the border is a bit hard to see, especially on white application backgrounds.
 
> Do we want to fix both of these within this PR? The 2nd issue could be fixed with using an (additional?) wlr_scene_rect for the background.

Hmm... For me the outline with white/black/white colors is recognizable enough even on white backgrounds. You can increase `resize.outlines.thickness` to make it more recognizable.

Of course we can add `wlr_scene_rect`, but then we need to repeat what we did when implementing snapping overlay: adding `resize.overlay.bg.{enabled,color}` and `resize.overlay.border.{enabled,color,width}`, and disabling `wlr_scene_rect` for pixman renderer. Also, it's not what remote desktop users would want. So I'm a bit less motivated for it.

**Comment by johanmalm** [comment]
LGTM.
Won't merge as there is a '[fixup]' commit still to be rebased at the end.

**Comment by tokyo4j** [comment]
I removed theme entries as I'm not 100% sure that the current approach with outlines is right.
Let's discuss about it later when users request settings for it or when we find a better approach.

This PR is good to go now. Sorry for my indecisiveness.

**Comment by johanmalm** [comment]
Sounds sensible. I've often gone for as few config/theme options as possible to start with. Can always add later as you say.

Thanks for PR. It's been on the cards for a while. Nice to get this in.

**Comment by Consolatis** [comment]
> Of course we can add `wlr_scene_rect`, but then we need to repeat what we did when implementing snapping overlay: adding `resize.overlay.bg.{enabled,color}` and `resize.overlay.border.{enabled,color,width}`, and disabling `wlr_scene_rect` for pixman renderer. Also, it's not what remote desktop users would want. So I'm a bit less motivated for it.

wlr_scene_rects should be no issue for pixman, its only if they are combined with transparency which causes visible performance degradation, at least from my experience. Fine with delaying this though. Thanks for the PR :)

**Comment by tokyo4j** [comment]
> wlr_scene_rects should be no issue for pixman, its only if they are combined with transparency which causes visible performance degradation, at least from my experience. Fine with delaying this though. Thanks for the PR :)

I thought showing resizing overlay with opaque color which hides the window contents is unacceptable, but I tried it and found it's not that bad (it flickers when finishing resizing though):

https://github.com/labwc/labwc/assets/22029524/69176d35-448d-424d-b783-237fcd508650

Perhaps this also works nicely with compression algorithm for remote desktop user.

Let's discuss it later anyway.

**Comment by Consolatis** [comment]
Technically, we could likely disable `view->scene_node` and delay disabling the overlay until the client commits a new size where we also enable `view->scene_node` again. That should get rid of the flicker and make the overlay a replacement for the surface itself.

**Pull request merged**